### ISSUE TITLE
[HOLD for payment 2023-08-17] [$1000] Web - invited members does not reflect in offline mode

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21891]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create a workspace if not created
2. Go to offline
2. Go to Settings > Workspaces > Members
3. Click on invite button
4. Enter email which is not in list and invite that member
5. Tt will navigates to Members page
6. Check for member which you invited

## Expected Result:
invited member should reflects in members list

## Actual Result:
invited member does not reflect in offline mode

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.33-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/896e0443-8036-4adb-a89b-f25f2af690b4

https://github.com/Expensify/App/assets/93399543/fbb21aa0-e27f-4216-bb43-e1106af843b6

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687960765759859

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01505e84ab8e8fed17</li>
        <li>Upwork Job ID: 1674841983137677312</li>
        <li>Last Price Increase: 2023-06-30</li>
    </ul>
</details>

**Comment by lschurr** [comment]
I'm not able to reproduce this. In offline mode, I invited a new user and could see them in the list.

**Comment by lschurr** [comment]
Bumped the Slack thread.

**Comment by rezkiy37** [comment]
Hi, I'm Michael (Mykhailo) from Callstack and I would like to work in this issue.

**Comment by rezkiy37** [comment]
It is reproducible. 
I'd suggest to build an optimistic object that includes personal details. It is similar when users create a new group when they are offline. 

<details><summary>Example</summary>

https://github.com/Expensify/App/assets/57314631/de608f6b-f5c1-45a0-9ff5-a0639ff1f85f

<img width="645" alt="Screenshot 2023-07-04 at 17 56 42" src="https://github.com/Expensify/App/assets/57314631/4b362b3f-b62a-4890-89e0-ed48c4019726">
</details>

**Comment by rezkiy37** [comment]
Figuring our how to merge a local created invited user with one who comes from BE.

**Comment by lschurr** [comment]
Not overdue.

**Comment by rezkiy37** [comment]
**Daily updates:**

Working on a solution how to add a new member to a workspace. I faced a problem when the app generates local users, it conflicts with the same upcoming users (_but with other ids_) from the backend. Looks like the app should not store the local generated users when a response comes. 

Planning to test all possible cases tomorrow: offline+new user, offline+old user, offline+old users; online+new user, online+old user, online+old users.

**Comment by rezkiy37** [comment]
**Daily updates**:

1. Figured out a situation how to not conflict with local created and backend upcoming invited members.
2. Tested all cases mentioned above.
3. A proposal in on internal review now.

**Comment by rezkiy37** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Initially, the problem occurs when the users are offline. They try to add new members to a workspace (_policy_). The entire flow works fine, except the result on UI.
When they invite new users (_the app hasn’t known about them before_), the app does not preview the pending state for them. It contradicts the app offline pattern - [Optimistic With Feedback Pattern B](https://github.com/Expensify/App/blob/main/contributingGuides/OFFLINE_UX.md#b---optimistic-with-feedback-pattern).

### What is the root cause of that problem?
Actually, the app does not create a base object for new users and does not set it to `personalDetails`. The app already sets these user **ids** to `policyMembers_{ID}`, but it includes the `accountID` and `pendingState` only. The workspace members page cannot preview, because of information lack.

### What changes do you think we should make in order to solve the problem?
The idea is to create a base object for the new users. The app should add these objects to `personalDetails`. It allows the app to use these information to preview all invited users.
However, once our users back online, the app sends a request to add new members. A response includes new members with own **ids**. It is important to clear local created members (_local **ids**_) and replace them with those members come from the backend.

**Changes:**
The app uses [**addMembersToWorkspace**](https://github.com/Expensify/App/blob/b105aa3d7776a164c476a982bf57939fa0ec34da/src/libs/actions/Policy.js#L355) function in **Policy.js** to execute this flow.
I am going to modify `memberList` of `successData` here. When an invited member already exists in `personalDetails`, the app should reset `pendingState` and `errors`. Otherwise, when an invited member is a new account, the app should remove it from the storage entirely, set `null`. This way after a successful response `memberList` does not has the redundant local created `accountID`.

Also, **addMembersToWorkspace** uses [**createPolicyExpenseChats**](https://github.com/Expensify/App/blob/b105aa3d7776a164c476a982bf57939fa0ec34da/src/libs/actions/Policy.js#L251) to build other request configs.
I am going to add `optimisticData` and `successData` for `ONYXKEYS.PERSONAL_DETAILS_LIST` (_`personalDetails`_).
On each iteration of `invitedEmailsToAccountIDs`, the app checks does this account already exist. **If not**,
1. Create a base object of a user and put it inside `optimisticData`. This way, during the offline or long time of waiting a response, the app could preview all invited members.
2. Clear that local create user in `successData` by setting `null`. This way I do not store redundant user after a response. I remove a copy of an upcoming backend user.

<details><summary>Example</summary>
<p>

Basically, I invite 2 users to a workspace in the offline mode. The first one is already created by the BE, the second one is being generated by the app locally.

**Part 1**

https://github.com/Expensify/App/assets/57314631/d8b20117-74b8-415f-8e4d-6d4a8503ecf1

**Part 2**

https://github.com/Expensify/App/assets/57314631/a7814c78-7f2d-4598-aca7-a6966c8d7379


</p>
</details> 

### What alternative solutions did you explore? (Optional)
N/A

**Comment by lschurr** [comment]
@0xmiroslav @rezkiy37 - any updates here?

**Comment by rezkiy37** [comment]
> [Proposal](https://github.com/Expensify/App/issues/21891#issuecomment-1628546127) looks good. I think it used to work before and this is recent regression. Maybe from secure logins migration refactor.
> 🎀 👀 🎀 C+ reviewed

Thank you for the review. Moving forward to the implementation 🫡

**Comment by rezkiy37** [comment]
I've opened a PR to the Callstack internal review. I will open it to the public review soon.

**Comment by rezkiy37** [comment]
Melvin, my proposal was approved. I am working on the issue.

**Comment by rezkiy37** [comment]
Could someone please add "Reviewing" label here?

**Comment by madmax330** [comment]
PR is in review

**Comment by rezkiy37** [comment]
The PR (https://github.com/Expensify/App/pull/22410) was approved.
https://github.com/Expensify/App/pull/22410#pullrequestreview-1562038130

**Comment by lschurr** [comment]
It looks like the automation didn't work on this one, but here's the payment summary:

Bug reporter: @gadhiyamanan $250
Contributor+: @situchan $1000

Could you confirm that's correct @madmax330?

**Comment by lschurr** [comment]
Looking at the PR, does @0xmiroslav also need to be paid for reviewing? https://github.com/Expensify/App/pull/22410

**Comment by situchan** [comment]
> Looking at the PR, does @0xmiroslav also need to be paid for reviewing? #22410

yes, they did the initial review

**Comment by lschurr** [comment]
Ok, so new payment summary:

- Bug reporter: @gadhiyamanan $250
- Contributor+: @0xmiroslav $1000
- Contributor+: @situchan $1000

**Comment by lschurr** [comment]
@0xmiroslav could you work through the checklist and confirm whether we need a regression test? https://github.com/Expensify/App/issues/21891#issuecomment-1673830490

**Comment by lschurr** [comment]
Payment for @0xmiroslav and @gadhiyamanan has been made in Upwork. 

@situchan could you accept the invite I sent and apply for the job? https://www.upwork.com/jobs/~01505e84ab8e8fed17

**Comment by situchan** [comment]
It seems invite is expired. I have no option to accept.

**Comment by lschurr** [comment]
Add paid. Closing :)
